soc/riscv/riscv-privilege/common/vector.S

```diff
@@ -25,12 +25,34 @@ SECTION_FUNC(vectors, __start)
 
 	.option norvc;
 
+#if defined(CONFIG_RISCV_MTVEC_VECTORED_MODE)
+    /*
+     * Set mtvec (Machine Trap-Vector Base-Address Register)
+     * to _ivt (interrupt vector table)
+     */
+    la t0, __ivt /* Load address of interrupt vector table */
+    addi t0, t0, 1 /* Enable vectored mode */
```

Comment:
I'm not sure what is happening here. From what I understand, vectored mode is enabled by writing 1 to the LSB of `mtvec`:
1. why are we writing to the vector table itself?
2. why adding - what if by some chance LSB is non-zero?
3. why aren't we preserving the other bits?

Reply:
Per the RISC-V privilege specification (v1.10 Section 3.1.12), mtvec has strict alignment rules. At at minimum, the vector table must be at an address that is 4-byte aligned (2 least significant bits are zero). In the most conservative case, the vector table must be 256-byte aligned. 

The alignment is required such that the LSB can be used to indicate which mode `mtvec` is in. For similar reasons as you pointed out above, if the vector table doesn't follow the alignment rules, then an unaligned address might indicate an unintended operating mode. When an interrupt occurs, the processor will mask, at minimum, the last 2 bits of the address stored in `mtvec`. If the result is 0x0, then the `mtvec` will jump to the address stored in `mtvec` mask with `0xFFFF_FFFC`, again at a minimum. If `mtvec` is in vectored mode, it jumps to the same address as in direct mode but with an added offset of `(mcause & 0x7FFF_FFFF) * 4`. The full equation for the vector handler address is given by `(mtvec & 0xFFFF_FFFC) + ((mcause & 0x7FFF_FFFF) * 4)`

In this implementation, I've forced the alignment of `__ivt` to 256-byte (line 54 of vector.S). I load that address into register `t0` via `la` instruction. `mtvec` is placed in vectored mode by adding 1 to the address of __ivt. For example, if `__ivt` is at address `0x2000_4100`, then the value stored in `mtvec` is `0x2000_4101`.  As described above, the processor will see that the LSB is a 1 indicated that the interrupt should be handled in vectored mode. The processor will jump to base address of `__ivt  + 4 x mcause` because it masks the LSB out of the address.

I can add explicit instructions to clear the LSB of `__ivt` prior to adding a one to the address so it's more explicit. It shouldn't be necessary though as the linker should follow the alignment rules specified. I can also add a condensed version of the explanation above into the code so that it's more clear in the code.

Please let me know if anything isn't clear.

Reply:
Thanks for you detailed explanation and for expanding the comments in your code. I think you can mark this conversation (and previous ones) as resolved.